### PR TITLE
Fix compiler and test failures with latest version of sigs.k8s.io/controller-runtime

### LINKED ISSUE #7882: Test Failures with latest version of sigs.k8s.io/controller-runtime
### Steps to reproduce

1. Checkout branch of latest dependabot PR with updates to sigs.k8s.io/controller-runtime eg. https://github.com/radius-project/radius/pull/7878
2. run 
```
           go get -u sigs.k8s.io/controller-runtime@latest
           go get -u k8s.io/client-go@latest
```
  ensure go.mod is updated with sigs.k8s.io/controller-runtime v0.19.0 and k8s.io/client-go v0.31.0   
4. run go mod tidy
5. run make build. Update code for errors (remove field `WarningHandler` as indicated in build errors):
```
        pkg/ucp/dataprovider/factory.go:67:3: unknown field WarningHandler in struct literal of type client.Options
        pkg/ucp/dataprovider/factory.go:67:33: undefined: runtimeclient.WarningHandlerOptions   
```
7. make build should now run fine.
8. run make test. this should return errors similar to for ~9 unit tests testing reconciler:
```
=== FAIL: pkg/controller/reconciler Test_DeploymentReconciler_ChangeEnvironmentAndApplication (0.00s)
    deployment_reconciler_test.go:81: 
                Error Trace:    ../pkg/controller/reconciler/deployment_reconciler_test.go:81
                                                        ../pkg/controller/reconciler/deployment_reconciler_test.go:142
                Error:          Received unexpected error:
                                controller with name deployment already exists. Controller names must be unique to avoid multiple controllers reporting to the same metric
                Test:           Test_DeploymentReconciler_ChangeEnvironmentAndApplication

=== FAIL: pkg/controller/reconciler Test_DeploymentReconciler_RadiusEnabled_ThenRadiusDisabled (0.00s)
    deployment_reconciler_test.go:81: 
                Error Trace:    ../pkg/controller/reconciler/deployment_reconciler_test.go:81
                                                        ../pkg/controller/reconciler/deployment_reconciler_test.go:216
                Error:          Received unexpected error:
                                controller with name deployment already exists. Controller names must be unique to avoid multiple controllers reporting to the same metric
                Test:           Test_DeploymentReconciler_RadiusEnabled_ThenRadiusDisabled
```

### Observed behavior

Unit Tests for controller failing with latest version of sigs.k8s.io/controller-runtime and related dependencies

### Desired behavior

Unit tests should not fail

### Workaround

Currently, we have not yet updated the codebase to the latest version of sigs.k8s.io/controller-runtime and it's related dependencies.

### rad Version

The latest version of the envtest tool used for unit testing in the controller-runtime project appears to be v0.17.0. We need to investigate if there is an incompatibility issue or if a code update is needed to ensure unit tests run successfully.

### Operating system

_No response_

### Additional context

_No response_

### Would you like to support us?

- [ ] Yes, I would like to support you

[AB#13145](https://dev.azure.com/azure-octo/e61041b4-555f-47ae-95b2-4f8ab480ea57/_workitems/edit/13145)

### PR DESCRIPTION
**Pull request by brooke-hamilton**
# Description

A [recent change](https://github.com/kubernetes-sigs/controller-runtime/pull/2903) to the [`controller-runtime`](https://github.com/kubernetes-sigs/controller-runtime/) package caused Radius compilation errors and test failures, as referenced in #7882. Breaking changes in `controller-runtime` include the removal of a configuration option, and a new validation that prevents duplicate controller names. This PR makes changes to update the packages referenced, fix the compilation errors, and address the test failures. 

Changes include:
- Updates to `go.mod` and `go.sum`.
- Fixed compiler errors by moving the warning suppression configuration from the `controller-runtime/pkg/client` options to `client-go/rest/Config`, as required by the changes in `controller-runtime`.
- Fixed broken tests by adding the `SkipNameValidation` configuration parameter to unit tests for creating new controllers.
- Some refactoring of commonly shared test functions and constants in `cli/controller/reconciler` into `shared-test.go`.

> NOTE: The compilation fix in this PR did not change the behavior of the `rad` cli--only the tests were changed. The assumption is that `rad` does not ever create controllers with duplicate names, so suppressing the duplicate controller name validation is not necessary, even though it is necessary in the automated tests.

## Type of change

- This pull request fixes a bug in Radius and has an approved issue (issue link required).

Fixes: #7882 

## Contributor checklist
Please verify that the PR meets the following requirements, where applicable:

- [ ] An overview of proposed schema changes is included in a linked GitHub issue.
- [ ] A design document PR is created in the [design-notes repository](https://github.com/radius-project/design-notes/), if new APIs are being introduced.
- [ ] If applicable, design document has been reviewed and approved by Radius maintainers/approvers.
- [ ] A PR for the [samples repository](https://github.com/radius-project/samples) is created, if existing samples are affected by the changes in this PR.
- [ ] A PR for the [documentation repository](https://github.com/radius-project/docs) is created, if the changes in this PR affect the documentation or any user facing updates are made.
- [ ] A PR for the [recipes repository](https://github.com/radius-project/recipes) is created, if existing recipes are affected by the changes in this PR.

### CODE REVIEW

File: pkg/controller/reconciler/deployment_reconciler_test.go [hunk 1]
```diff
@@ -74,7 +77,7 @@ func SetupDeploymentTest(t *testing.T) (*mockRadiusClient, client.Client) {
 	err = (&DeploymentReconciler{
 		Client:        mgr.GetClient(),
 		Scheme:        mgr.GetScheme(),
-		EventRecorder: mgr.GetEventRecorderFor("recipe-controller"),
+		EventRecorder: mgr.GetEventRecorderFor("deployment-controller"),
```
Comment (rynowak) [inline pkg/controller/reconciler/deployment_reconciler_test.go:80]: 👍

**Comment by rynowak** [comment]
Looks like `make generate` needs to run again. Your other changes probably updated a code generator we're using for the K8s controllers.

> Fixed broken tests by adding the SkipNameValidation configuration parameter to unit tests for creating new controllers.

Do you know if `SkipNameValidation` is a new kind of warning they added to controller-runtime?

I agree with suppressing the warning. It's probably being triggered because our tests will register a new instance of the same controller for each test.

**Review by rynowak — COMMENTED** [review]
Changes in this PR so far look right 👍. Will look again after you resolve build issues.

**Comment by brooke-hamilton** [comment]
> Looks like `make generate` needs to run again. Your other changes probably updated a code generator we're using for the K8s controllers.
> 
> > Fixed broken tests by adding the SkipNameValidation configuration parameter to unit tests for creating new controllers.
> 
> Do you know if `SkipNameValidation` is a new kind of warning they added to controller-runtime?
> 
> I agree with suppressing the warning. It's probably being triggered because our tests will register a new instance of the same controller for each test.

I'm having an issue with the `make generate-controller` target. I will follow up.

[Validation](https://github.com/kubernetes-sigs/controller-runtime/blob/4381fa0aeee43e331be14b0d70cd276e1e91ad7a/pkg/controller/name.go#L29) to prevent duplication of controller names was added in [PR2902](https://github.com/kubernetes-sigs/controller-runtime/pull/2902), and a few days later [PR2918](https://github.com/kubernetes-sigs/controller-runtime/pull/2918) added the [`SkipNameValidation`](https://github.com/kubernetes-sigs/controller-runtime/blob/4381fa0aeee43e331be14b0d70cd276e1e91ad7a/pkg/controller/controller.go#L153) configuration option which bypasses the validation for duplicate names.

**Comment by rynowak** [comment]
Thanks. Based on this comment, we're in exactly the same test scenario as the user here: https://github.com/kubernetes-sigs/controller-runtime/pull/2902#issuecomment-2284080131

Sounds like it's intended we use `SkipNameValidation`.

**Comment by brooke-hamilton** [comment]
@rynowak I needed to bump the `controller-gen` version to align with the current version of k8s.io/*. https://github.com/kubernetes-sigs/controller-tools

File: pkg/controller/reconciler/recipe_reconciler_test.go [hunk 2]
```diff
@@ -56,7 +48,10 @@ func SetupRecipeTest(t *testing.T) (*mockRadiusClient, client.Client) {
 	t.Cleanup(cancel)
 
 	mgr, err := ctrl.NewManager(config, ctrl.Options{
-		Scheme: scheme,
+		Scheme: scheme, 
+		Controller: crconfig.Controller{
```
Comment (lakshmimsft) [inline pkg/controller/reconciler/recipe_reconciler_test.go:52]: 👍

File: pkg/controller/reconciler/recipe_reconciler_test.go [hunk 3]
```diff
@@ -56,7 +48,10 @@ func SetupRecipeTest(t *testing.T) (*mockRadiusClient, client.Client) {
...
+			SkipNameValidation: boolPtr(true),
```
Comment (lakshmimsft) [inline pkg/controller/reconciler/recipe_reconciler_test.go:53]: 👍

File: pkg/ucp/queue/provider/factory.go [hunk 4]
```diff
@@ -63,13 +67,9 @@ func initAPIServer(ctx context.Context, opt QueueProviderOptions) (queue.Client,
...
 	options := runtimeclient.Options{
 		Scheme: scheme,
-
-		// The client will log info the console that we don't really care about.
-		WarningHandler: runtimeclient.WarningHandlerOptions{
-			SuppressWarnings: true,
-		},
 	}
 
+
```
Comment (lakshmimsft) [inline pkg/ucp/queue/provider/factory.go:72]: nit: extra line

Comment (brooke-hamilton) [reply pkg/ucp/queue/provider/factory.go:72]: Thank you!

**Review by lakshmimsft — APPROVED** [review]

**Pull request merged**